cpmpy/solvers/ortools.py

```diff
@@ -207,7 +207,9 @@ def solveAll(self, display=None, time_limit=None, solution_limit=None, **kwargs)
 
             Returns: number of solutions found
         """
-        # XXX: check that no objective function??
+        if self.ort_model.HasObjective():
+            raise Exception("OR-tools does not support finding all optimal solutions.")
```

Comment:
I think these two lines can go into the template... by default solvers will not support finding all optimal solutions... also the generic one does not (and should not)

I am not convinced of your solver_interface generic implementation either, you should add a constraint that the subsequetn solutions ahve to be better than the optimal one after finding it. But I don't think we should go there... we can just stick to 'solveAll is for satisfaction problems'... and let the user model it manually otherwise
(though I am fine with leaving it in for gurobi now that you alrady implemetned it, we just raise for all other solvers)

Reply:
Why should we add a constraint forcing the objective to be better? We just walk over the different optimal solutions by adding no-goods as usual?

Reply:
That is what you do now, and then you break when a sub-optimal solution is found.

But in this case we know that the solution needs at least/most a certain value, so it does not need to search for that lower/upper bound each time again...

I don't have use cases, but I would assume that it would be faster to give the solver more information.
At the very least, in the very last call (that would be suboptimal), it can stop as soon as its lb/ub is below the optimal value, without having to prove that whatever value it then has is the now optimal one...

Reply:
Yes, but if we implement it as we did in `solver_interface` the solver will here also become invalid... But as OR-tools is not incremental i guess that doesn't really matter here. I'll implement it and raise a warning as in `solver_interface`

Reply:
Aha, but then we need to store the objective function in the solver object as well. I don't see any harm in doing so, but seems silly to keep it in there just for this (obscure) feature?

Reply:
Ok will leave it for now and raise error when calling solveAll with objective.